### ISSUE TITLE
PublishedServerUrl does not function correctly

### ISSUE THREAD

**Issue by dbaker-rh** [jellyfin/jellyfin#5280]
**Describe the bug**

The "PublishedServerUrl" feature does not function correctly.

Tested on docker.io/jellyfin/jellyfin:10.7.0-rc3 as well as :unstable and :latest

Part 1:

If I pass in the environment variable JELLYFIN_PublishedServerUrl it appears to be ignored.

Passing the --published-server-url on the command line does work.

Part 2:

However, even when passing on the command line, the port number is ignored and the port of the jellyfin listener is published, not the port number from the parameter which is stripped and ignored.

**System (please complete the following information):**
 - OS: [e.g. Debian, Windows]  kubernetes / k3s 1.20
 - Virtualization: [e.g. Docker, KVM, LXC]  kube
 - Clients: [Browser, Android, Fire Stick, etc.]  android, web
 - Browser: [e.g. Firefox 72, Chrome 80, Safari 13]  any
 - Jellyfin Version: [e.g. 10.4.3, nightly 20191231]  10.7.0-rc3, unstable, latest
 - Playback: [Direct Play, Remux, Direct Stream, Transcode]  via chromecast
 - Installed Plugins: [e.g. none, Fanart, Anime, etc.] none
 - Reverse Proxy: [e.g. none, nginx, apache, etc.] traefik v2
 - Base URL: [e.g. none, yes: /example] none
 - Networking: [e.g. Host, Bridge/NAT] kubernetes
 - Storage: [e.g. local, NFS, cloud] local

**To Reproduce**

Run jellyfin in kubernetes.  Pass http://jellyfin.local/ through to the pod on default port 8096

Run a sidecar tcpdump container to monitor traffic.

With no PublishedServerUrl (or passed as environment, not on command line)
  tcpdump -i eth0 -A | grep ^LOCATION
  LOCATION: http://10.42.0.210:8096/dlna/1fb3e71a56654e22a80dc88faf6079f4/description.xml

With --published-server-url http://jellyfin.local/
  tcpdump -i eth0 -A | grep ^LOCATION
  LOCATION: http://jellyfin.local:8096/dlna/1fb3e71a56654e22a80dc88faf6079f4/description.xml

With --published-server-url http://jellyfin.local:80/
  tcpdump -i eth0 -A | grep ^LOCATION
  LOCATION: http://jellyfin.local:8096/dlna/1fb3e71a56654e22a80dc88faf6079f4/description.xml

(note port numbers in the last two examples)

**Expected behavior**

Part A - published server URL should be recognized from environment, as documentation shows
Part B - published server URL should have it's port number included when publishing the LOCATION: strings

**Logs**

n/a

**Screenshots**

n/a

**Additional context**

Fixing this may be one of the few final steps to getting chromecast integration working perfectly from within kube (and by extension, other docker) environments

I have shrunk my kubernetes cluster to a single node for diagnostics.  In this state, I can work around the problem in the short term by turning tcp/8096 into a hostPort for the pod.  Definitely far from ideal.

**Comment by dbaker-rh** [comment]
This looks to be related to these lines:   https://github.com/jellyfin/jellyfin/blob/master/Emby.Dlna/Main/DlnaEntryPoint.cs#L315-L318

                var uri = new UriBuilder(_appHost.GetSmartApiUrl(address.Address) + descriptorUri);
                // DLNA will only work over http, so we must reset to http:// : {port}
                uri.Scheme = "http://";
                uri.Port = _netConfig.HttpServerPortNumber;


The GetSmartApiUrl looks to pull from PublishedServerUrl if it's defined, but line 318 overwrites the port of the local listener, not the port if it's defined in the PublishedServerUrl


I rebuilt the container with a very small code diff against *release-10.7.z* to hardcode my desired port 80.

````
$ git diff
diff --git a/Emby.Dlna/Main/DlnaEntryPoint.cs b/Emby.Dlna/Main/DlnaEntryPoint.cs
index 82490ec31..678568ca8 100644
--- a/Emby.Dlna/Main/DlnaEntryPoint.cs
+++ b/Emby.Dlna/Main/DlnaEntryPoint.cs
@@ -315,7 +315,7 @@ namespace Emby.Dlna.Main
                 var uri = new UriBuilder(_appHost.GetSmartApiUrl(address.Address) + descriptorUri);
                 // DLNA will only work over http, so we must reset to http:// : {port}
                 uri.Scheme = "http://";
-                uri.Port = _netConfig.HttpServerPortNumber;
+                uri.Port = 80;
 
                 var device = new SsdpRootDevice
                 {

````

Now, my tcpdump shows this

````
LOCATION: http://jellyfin.local/1fb3e71a56654e22a80dc88faf6079f4/description.xml
````


Chromecast is still not stable, but this looks to be one step towards that goal.



TLDR - if the configuration specifies a PublishedServerUrl I think you should believe it and use it as-is.  Separately, a warning if it specifies https would be more appropriate than just forcefully stripping the user provided value and replacing it.

**Comment by BaronGreenback** [comment]
Just to clarify: 

You have port 80 redirected to 8096 internally, so expect the dlna to publish http://yyyy not http:/yyyy:8096?

**Comment by dbaker-rh** [comment]
Yes.  I have jellyfin running inside a kubernetes cluster.  My ingress actually takes http://yyyy and redirects to https://yyyy.  It also takes https://yyyy/, and redirects it in plaintext to the pod inside running on port 8096.

Functionally, it looks like as you describe.


With that one change I now actually have full Chromecast functionality from the android app, which seems to be a long sought and struggled for feature based on looking at past issues and forum posts.

I'm happy share my yaml configs if you want to recreate the environment.

**Comment by BaronGreenback** [comment]
First PR https://github.com/jellyfin/jellyfin/pull/5312

**Comment by BaronGreenback** [comment]
@dbaker-rh Any chance you could test PR https://github.com/jellyfin/jellyfin/pull/5315 as it should make JELLYFIN_PublishedServerUrl functional again?

**Comment by BaronGreenback** [comment]
@dbaker-rh Any chance you could test PR https://github.com/jellyfin/jellyfin/pull/5315 as it should make JELLYFIN_PublishedServerUrl functional again?

**Comment by dbaker-rh** [comment]
No problem - I'll poke at it tomorrow.

**Comment by dbaker-rh** [comment]
#5312
Change seems to work well.
I would consider adding a small comment since the behaviour is a little non obvious.  Perhaps, "If publishedserverurl is given use it as-is, otherwise try to construct a sane default"

#5315 
I noted the comment in that PR regarding unexpected overriding of other values.  Aside from that concern, I can confirm that JELLYFIN_PublishedServerUrl works as an env variable again.


Thanks.
